cmd/thanos/receive.go

```diff
@@ -214,235 +214,302 @@ func runReceive(
+				return errors.Wrap(webHandler.Run(), "error starting web server")
+			},
+			func(err error) {
+				webHandler.Close()
+			},
+		)
+	}
+
+	level.Info(logger).Log("msg", "starting receiver")
+	return nil
+}
+
+func setupGRPCServer(g *run.Group,
+	logger log.Logger,
+	reg *prometheus.Registry,
+	tracer opentracing.Tracer,
+	conf *receiveConfig, s *grpcserver.Server,
+	startGRPC chan struct{},
+	dbReady chan struct{},
+	comp component.SourceStoreAPI,
+	dbs *receive.MultiTSDB,
+	webHandler *receive.Handler,
+	grpcLogOpts []grpc_logging.Option,
+	tagOpts []tags.Option,
+	grpcProbe *prober.GRPCProbe,
+
+) error {
+	g.Add(func() error {
+		defer close(startGRPC)
+
+		tlsCfg, err := tls.NewServerConfig(log.With(logger, "protocol", "gRPC"), *conf.grpcCert, *conf.grpcKey, *conf.grpcClientCA)
+		if err != nil {
+			return errors.Wrap(err, "setup gRPC server")
+		}
+
+		for range dbReady {
+			if s != nil {
+				s.Shutdown(errors.New("reload hashrings"))
 			}
 
-			// Check the hashring configuration on before running the watcher.
-			if err := cw.ValidateConfig(); err != nil {
-				cw.Stop()
-				close(updates)
-				return errors.Wrap(err, "failed to validate hashring configuration file")
+			rw := store.ReadWriteTSDBStore{
+				StoreServer: store.NewMultiTSDBStore(
+					logger,
+					reg,
+					comp,
+					dbs.TSDBStores,
+				),
+				WriteableStoreServer: webHandler,
 			}
 
-			ctx, cancel := context.WithCancel(context.Background())
-			g.Add(func() error {
-				level.Info(logger).Log("msg", "the hashring initialized with config watcher.")
-				return receive.HashringFromConfigWatcher(ctx, updates, cw)
-			}, func(error) {
-				cancel()
-			})
-		} else {
-			var ring receive.Hashring
-			// The Hashrings config file content given initialize configuration from content.
-			if len(conf.hashringsFileContent) > 0 {
-				ring, err = receive.HashringFromConfig(conf.hashringsFileContent)
-				if err != nil {
-					close(updates)
-					return errors.Wrap(err, "failed to validate hashring configuration file")
-				}
-				level.Info(logger).Log("msg", "the hashring initialized directly with the given content through the flag.")
-			} else {
-				level.Info(logger).Log("msg", "the hashring file is not specified use single node hashring.")
-				ring = receive.SingleNodeHashring(conf.endpoint)
+			s = grpcserver.New(logger, &receive.UnRegisterer{Registerer: reg}, tracer, grpcLogOpts, tagOpts, comp, grpcProbe,
+				grpcserver.WithServer(store.RegisterStoreServer(rw)),
+				grpcserver.WithServer(store.RegisterWritableStoreServer(rw)),
+				grpcserver.WithListen(*conf.grpcBindAddr),
+				grpcserver.WithGracePeriod(time.Duration(*conf.grpcGracePeriod)),
+				grpcserver.WithTLSConfig(tlsCfg),
+			)
+			startGRPC <- struct{}{}
+		}
+		if s != nil {
+			s.Shutdown(err)
+		}
+		return nil
+	}, func(error) {})
+	// We need to be able to start and stop the gRPC server
+	// whenever the DB changes, thus it needs its own run group.
+	g.Add(func() error {
```

Comment:
The function name says `setupGRPCServer` but this also starts servers. It'd be good to scope of functions small as possible.

Reply:
Shall we split up into two small functions?

Reply:
Done

Reply:
For some reason, splitting up tasks are creating a nil pointer for the grpcServer, so I reverted back the change. Let me know if you have any suggestion